src/eval.c

```diff
@@ -3665,6 +3673,350 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    ga_clear(&current);
+    return OK;
+}
+
+/*
+ * Refreshes previous_quote with NUL
+ *	if the closing quote of
+ *	the opening quote (previous_quote) found on **source.
+ * Refreshes previous_quote with ' or "
+ *	if a opening quote is never found and that quote found.
+ * Doesn't Refresh if no quotes found.
+ */
+    static void
+refresh_pair_of_quote(char_u *previous_quote, char_u source)
+{
+    if (*previous_quote == NUL && source == '\'')
+    {
+	*previous_quote = '\'';
+	return;
+    }
+    else if (*previous_quote == NUL && source == '"')
+    {
+	*previous_quote = '"';
+	return;
+    }
+
+    if (*previous_quote == '\'' && source == '\'')
+	*previous_quote = NUL;
+    else if (*previous_quote == '"' && source == '"')
+	*previous_quote = NUL;
+}
+
+    static int
+is_escaped_quote(int is_literal_string, char_u *p)
+{
+    return
+	(is_literal_string && *p == '\'' && *(p + 1) == '\'') ||
+	(!is_literal_string && *p == '\\' && *(p + 1) == '"');
+}
+
+/*
+ * Finds closing '}' of taken expr
+ */
+    static int
+forward_to_end_of_template_expr(char_u **expr, int is_literal_string)
+{
+    char_u  *expr_head = *expr;
+    size_t  nested_blocks = 0;
+    char_u  encountered_quote = NUL;
+
+    // While the closing '}' encountered. The '}' is neither '}' of a dict nor
+    // '}' in a string.
+    for (; !(encountered_quote == NUL && nested_blocks == 0 && **expr == '}');
+							MB_PTR_ADV(*expr))
+    {
+	switch (**expr)
+	{
+	    case '{':
+		if (encountered_quote == NUL)
+		    ++nested_blocks;
+		break;
+	    case '}':
+		if (encountered_quote == NUL)
+		    --nested_blocks;
+		break;
+	    case NUL:
+		semsg(_("E451: Unterminated template expression: %s"), expr_head);
+		return FAIL;
+	    default:
+		if (is_escaped_quote(encountered_quote == '\'', *expr))
+		{
+		    ++*expr;
+		    break;
+		}
+		refresh_pair_of_quote(&encountered_quote, **expr);
+		break;
+	}
+    }
+
+    if (*expr == expr_head)
+    {
+	semsg(_("E452: Empty template expression: %s"), expr_head);
+	return FAIL;
+    }
+
+    return OK;
+}
+
+/*
+ * Get expr of "${" expr  "}" of source,
+ * and forward source.
+ */
+    static char_u*
+read_template_expr(char_u **source, int is_literal_string)
+{
+    char_u  *expr_head;
+
+    *source += 2;  // forward with beginning '${'
+    expr_head = *source;
+
+    if (!forward_to_end_of_template_expr(source, is_literal_string))
+	return NULL;
+
+    return vim_strnsave(expr_head, (int)(*source - expr_head));
+}
+
+    static char_u*
+stringify_expr(char_u *expr)
+{
+    typval_T result, string_result;
+
+    if (eval1(&expr, &result, TRUE) == FAIL)
+	return NULL;
+
+    // If $UNDEFINED_ENV_VAR evaluated
+    if (result.v_type != VAR_STRING)
+    {
+	f_string(&result, &string_result);
+	clear_tv(&result);
+	result = string_result;
+    }
+    else if (result.vval.v_string == NULL)
+    {
+	result.vval.v_string = (char_u *) alloc(1);  // meaning success of this function
+	result.vval.v_string[0] = '\0';
+    }
```

Comment:
This can be a bit simpler:

```suggestion
	result.vval.v_string = ALLOC_CLEAR_ONE(char_u);  // meaning success of this function
```

Reply:
thx!